Write Batchv1JobLifecycleTest + 4 Endpoints

riaankleinhans opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- patchBatchV1NamespacedJob
- replaceBatchV1NamespacedJob
- deleteBatchV1CollectionNamespacedJob
- listBatchV1JobForAllNamespaces

**Which issue(s) this PR fixes:**
Fixes #108641

**Testgrid Link:**
[Testgrid](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20manage%20the%20lifecycle%20of%20a%20job&graph-metrics=test-duration-minutes)


**Special notes for your reviewer:**
Adds +4 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

Riaankl commented:
/triage accepted

alculquicondor commented:
wonderful!
Let me add it to my list for early next week :)

/assign

Riaankl commented:
/assign @soltysh @kow3ns

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
+
+	ginkgo.It("should test the lifecycle of a job", func() {
```
Can you add a summary of the test as a comment above?

heyste replied:
Added a description for the test. I've used the [test comment metadata](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-comment-metadata) format required for conformance tests. Saving some time for when the test is [promoted to conformance](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#promoting-tests-to-conformance) in the near future.

alculquicondor replied:
Oh, same problem here

```suggestion
	ginkgo.It("should manage the lifecycle of a job", func() {
```

`It` refers to the API and controller, Job in this case, not the test.

heyste replied:
Fixed

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", ns)
+
+		ginkgo.By("Ensuring pods aren't created for job")
+		framework.ExpectEqual(wait.Poll(framework.Poll, wait.ForeverTestTimeout, func() (bool, error) {
+			pods, err := e2ejob.GetJobPods(f.ClientSet, ns, job.Name)
+			if err != nil {
+				return false, err
+			}
+			return len(pods.Items) > 0, nil
+		}), wait.ErrWaitTimeout)
```
This kind of wait (wait for nothing to happen) is expensive and is already tested in another test. Do you really need to test it again here?

heyste replied:
Agree. Removed

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+		framework.ExpectNoError(err, "failed to retrieve latest job object")
+		exists := false
+		for _, c := range job.Status.Conditions {
+			if c.Type == batchv1.JobSuspended {
```
I would just check that this condition showed up, instead of checking the pods not to appear.

heyste replied:
Agree. Removed

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+			},
+		}
+		jobsList, err := jobClient.List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
+		framework.ExpectNoError(err, "failed to list Pods")
```
Maybe check that the list is empty?

heyste replied:
There are other list checks later in the test that provide more value.

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+		framework.ExpectEqual(exists, true)
+
+		ginkgo.By("Patching the Job")
+		patchJob, err := json.Marshal(v1.PodTemplateSpec{
```
why `PodTemplateSpec`. Shouldn't it be `Job`?

heyste replied:
The name `patchJob` was a bit confusing. I've renamed it to `payload`. From what I see in the documentation Job is using a [pod template](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-template) within the spec. From the e2e test logs it passes after finding the required label. Have I missed something?

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+		})
+		if err != nil {
+			p, _ := jobClient.Get(context.TODO(), jobName, metav1.GetOptions{})
+			framework.Logf("Job: %+v", p)
```
What's the intent of this log line?

heyste replied:
Provide more detail about the Job if the required event was not found. There have been a number of issues trying to confirm what has caused a test to flake when dealing with watch events.

alculquicondor replied:
Can you include some context in the message? Perhaps:

```
framework.Logf("Job still exists after deletion call: %+v", p)
```

Also, why is the variable named `p`? Should it be `j` for job?

heyste replied:
Slipped though from one of my other status tests. Fixed variable and message for each watch scenario.

heyste commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +368,172 @@ var _ = SIGDescribe("Job", func() {
+				Labels: map[string]string{jobName: "patched"},
+			},
+			Spec: v1.PodSpec{
+				TerminationGracePeriodSeconds: &one,
```
I picked this style up from another e2e test and looks to be the general style used in e2e tests

alculquicondor replied:
I would imagine some tests where written before the pointer package existed.

In general, defining constants for tests cases is not a very good practice as it makes the test harder to read.

heyste replied:
Fixed

heyste commented:
/test pull-kubernetes-integration
multiple unrelated flakes

```
no cloud provider provided
```

heyste commented:
/test pull-kubernetes-e2e-kind
unrelated flake

```
 ✓ Joining worker nodes 🚜
 • Waiting ≤ 1m0s for control-plane = Ready ⏳  ...
 ✗ Waiting ≤ 1m0s for control-plane = Ready ⏳
 • WARNING: Timed out waiting for Ready ⚠️ 
```

heyste commented:
/test pull-kubernetes-e2e-kind-ipv6
unrelated flake
```
 ✓ Joining worker nodes 🚜
 • Waiting ≤ 1m0s for control-plane = Ready ⏳  ...
 ✗ Waiting ≤ 1m0s for control-plane = Ready ⏳
 • WARNING: Timed out waiting for Ready ⚠️ 
```

heyste commented:
/hold 

There looks to be an issue patching a job. The following doesn't fail 

```
patchedJob, err := f.ClientSet.BatchV1().Jobs(ns).Patch(context.TODO(), jobName, types.StrategicMergePatchType, []byte(payload), metav1.PatchOptions{})
framework.ExpectNoError(err, "failed to patch Job %s in namespace %s", jobName, ns)
```
but... I can't locate the patched changes. An example of the extra logging produces
```
Mar 24 17:11:35.705: INFO: Labels: map[string]string{"e2e-job-label":"e2e-rzdth"}
Mar 24 17:11:35.705: INFO: Spec Template Labels: map[string]string{"controller-uid":"ad86d622-e2d3-454b-abdf-c59995e0793c", "job":"e2e-rzdth", "job-name":"e2e-rzdth"}
Mar 24 17:11:35.705: INFO: TerminationGracePeriodSeconds: 30
```
Which is missing a new label and `TerminationGracePeriodSeconds` should be 1

heyste commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +369,166 @@ var _ = SIGDescribe("Job", func() {
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", ns)
+
+		ginkgo.By("Patching the Job")
+		payload, err := json.Marshal(v1.PodTemplateSpec{
```
Switched to using `batchv1.JobSpec` which doesn't return an error but the resulting patch doesn't get applied from what I can see.

heyste replied:
The patch uses a small snippet of JSON that adds a label.

https://github.com/kubernetes/kubernetes/blob/e6d5c32ee4629f9e23238506ff2c28656d3844dc/test/e2e/apps/job.go#L531-L532

heyste commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +369,166 @@ var _ = SIGDescribe("Job", func() {
+		ginkgo.By("Watching for Job to be updated")
+		ctx, cancel = context.WithTimeout(context.Background(), f.Timeouts.PodStart)
+		defer cancel()
+		_, err = watchtools.Until(ctx, jobsList.ResourceVersion, w, func(event watch.Event) (bool, error) {
```
When the patch issue is resolved I'll revisit this request.

heyste replied:
Created a function that contains all the watch tracking / logging.

https://github.com/kubernetes/kubernetes/blob/e6d5c32ee4629f9e23238506ff2c28656d3844dc/test/e2e/apps/job.go#L617-L620

heyste committed:
Create job lifecycle e2e test

The test validates the following endpoints
- deleteBatchV1CollectionNamespacedJob
- listBatchV1JobForAllNamespaces
- patchBatchV1NamespacedJob
- replaceBatchV1NamespacedJob

heyste commented:
/unhold  

Fixed issue with patching the job. Using `batchv1.Job` was trying to change some immutable fields within the spec template.

heyste commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel
unrelated flake
```
: "timed out waiting for the condition",
```

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -48,6 +53,18 @@ import (
+	jobName         string
+	watchEvent      watch.EventType
+	extJob          *batchv1.Job
+	internalKind    string
+	internalValue   string
+	externalValue   string
```
these variable names are very confusing.

I would say:
```
updatedMetadataType
updatedKey
updatedValue
```

but I'm open to suggestions

heyste replied:
Fixed!. Thank for the suggested wording as I wasn't happy with it either but now we are! 💯

alculquicondor reviewed (commented):
/lgtm

soltysh reviewed (approved):
/approve
/priority backlog
